File: ruby_api/threat_recon_request.rb
#!/usr/bin/env ruby

require 'rest-client'
require 'json'

@url = "https://api.threatrecon.co:8080/api/v1/search"
api_key = 'your-api-key'
RestClient.proxy = ENV['https_proxy']

def query_threat_recon(indicator,api_key)
	response = RestClient.post @url, {'api_key' => api_key,'indicator' => indicator}
	case response.code
		when 200
			response_json = JSON.parse(response)
		else
			raise "Unknown Error"
	end
end

puts "Please Enter an Indicator: "
search = gets.chomp

results = query_threat_recon(search,api_key)

indicator_meta = Array.new
related_indicators = Array.new

if results["Msg"].eql?("Success")
	results["Results"].each do |item|
		if item["RootNode"].empty? and search.eql?(item["Indicator"])
			indicator_meta << {"Reference" => item["Reference"],
					  "Source" => item["Source"],
					  "KillChain" => item["KillChain"],
					  "FirstSeen" => item["FirstSeen"],
					  "LastSeen" => item["LastSeen"],
					  "Attribution" => item["Attribution"],
					  "ProcessType" => item["ProcessType"],
					  "Country" => item["Country"],
					  "Tags" => item["Tags"],
					  "Comments" => item["Comments"],
					  "Confidence" => item["Confidence"]}
		end
		if item["RootNode"].eql?(search)
			related_indicators << {"Indicator" => item["Indicator"],
                                       	       "ProcessType" => item["ProcessType"],
                                               "Rdata" => item["Rdata"],
                                               "Rrname" => item["Rrname"],
                                               "RootNode" => item["RootNode"]}
		end
		if item["RootNode"] != "" and search != item["RootNode"]
                        related_indicators << {"Indicator" => item["Indicator"],
                                               "ProcessType" => item["ProcessType"],
                                               "Rdata" => item["Rdata"],
                                               "Rrname" => item["Rrname"],
                                               "RootNode" => item["RootNode"]}		
		end
	end
end

if indicator_meta.length.eql?(0) and related_indicators.length.eql?(0)
	puts "COMMENT: #{search} is a derived indicator...metadata is inherited from the root node"
end

if indicator_meta.length > 0
	puts "ThreatRecon has found the following metadata on #{search}"
	indicator_meta.first.each_key do |key|
		puts "#{key} : #{indicator_meta.first[key]}"
	end
end

if related_indicators.length > 0
	puts "ThreatRecon has found the following indicator(s) that are related to #{search}"
	related_indicators.each do |elem|
		if elem["Indicator"] != search
			puts
			puts "******************"
			puts "Related indicator: #{elem["Indicator"]}"
			puts "Relationship Type: #{elem["ProcessType"]}"
			puts "Relationship Pivot: #{elem["Rdata"]}" unless elem["Rdata"].empty?
			puts "Rrname: #{elem["Rrname"]}" unless (elem["Rrname"].nil? or elem["Rrname"].empty?) and (elem["Rrname"] != elem["Indicator"])
			puts "******************"
		else
			                        puts
                        puts "******************"
                        puts "Related indicator: #{elem["RootNode"]}"
                        puts "Relationship Type: #{elem["ProcessType"]}"
                        puts "Relationship Pivot: #{elem["Rdata"]}" unless elem["Rdata"].empty?
                        puts "Rrname: #{elem["Rrname"]}" unless (elem["Rrname"].nil? or elem["Rrname"].empty?) and (elem["Rrname"] != elem["RootNode"])
                        puts "******************"

		end
	end
end

File: examples/threat_recon_request.py
"""
Copyright (C) 2014 by Wapack Labs Corporation
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.
You should have received a copy of the GNU General Public License
along with this program. If not, see <http://www.gnu.org/licenses/>.
"""


import socket
from threatrecon.api import get_api_key
from threatrecon.query import raw_query_threat_recon


api_key = get_api_key() or 'my api key'
search = raw_input("Please Enter an indicator: ")

results = raw_query_threat_recon(search, api_key)

indicator_meta = []
related_indicators = []


#check host IP if no results
if not results:     # results is an empty list
    try:    # tries to get IP from domain
        iplookup = socket.gethostbyname(search)
        print "***** No results found for this domain..."
        print "***** checking host IP: %s\n" % iplookup
        results = raw_query_threat_recon(iplookup, api_key)
    except socket.gaierror as e:
        print "***** Lookup failed: %s" % e

#find relationships in JSON results and list out
else:
    for item in results:
        root_node = item["rootnode"]
        rdata = item["rdata"]
        indicator = item["indicator"]
        if item["rootnode"] == '' and search == item["indicator"]:
            indicator_meta.append(
                [
                    item["reference"],
                    item["source"],
                    item["killchain"],
                    item["firstseen"],
                    item["lastseen"],
                    item["attribution"],
                    item["processtype"],
                    item["country"],
                    item["tags"],
                    item["comment"],
                    str(item["confidence"])
                ]
            )
        if search == item["rootnode"]:
            related_indicators.append(
                [
                    item["indicator"],
                    item["processtype"],
                    item["rdata"],
                    item["rrname"],
                    item["rootnode"]
                ]
            )
        if item["rootnode"] != '' and search != item["rootnode"]:
            related_indicators.append(
                [
                    item["indicator"],
                    item["processtype"],
                    item["rdata"],
                    item["rrname"],
                    item["rootnode"]
                ]
            )


#check to see if search is a derived indicator
if not indicator_meta and related_indicators:
    print "COMMENT: %s is a derived indicator..." % search
    print "metadata is inherited from the root node"

if indicator_meta:
    list_meta = indicator_meta[0]
else:
    list_meta = []

if not list_meta and not related_indicators:
    print 'sorry, no results.. that might be a good thing'

#list metadata if Direct indicator
if list_meta:
    print "Threat Recon has found the following metadata on %s: " % search
    tags = [
        'Reference',
        'Source',
        'KillChain',
        'First Seen',
        'Last Seen',
        'Attribution',
        'ProcessType',
        'Country',
        'TAGS',
        'Comment',
        'Confidence'
    ]
    for item in indicator_meta:
        for i in range(11):
            if item[i]:
                print "%s: %s" % (tags[i], item[i])
        print '\n'

#list related indicators
if related_indicators:
        print "\n"
        print "Threat Recon has found the following indicator(s) "
        print "that are related to %s" % search
        for item in related_indicators:
            if item[0] != search:
                print '******************'
                print 'Related indicator: %s' % item[0]
                print 'Relationship type: %s' % item[1]
                if item[2]:
                        print 'Relationship pivot: %s' % item[2]
                if item[3] and item[3] != item[0]:
                        print 'RRNAME: %s' % item[3]
                print '****************** \n'
            else:
                print '******************'
                print 'Related indicator: %s' % item[4]
                print 'Relationship type: %s' % item[1]
                if item[2]:
                        print 'Relationship pivot: %s' % item[2]
                if item[3] and item[3] != item[4]:
                        print 'RRNAME: %s' % item[3]
                print '****************** \n'
